1.0.0 updates

stephencelis opened the pull request:
Assorted fixes and enhancements. Comments, @kainosnoema?

stephencelis committed:
Ruby 2 syntax

Let's use keyword arguments. It'll make things simpler when we add more
to the Jsonite.property declaration.

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Remove unneeded parens

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Add Jsonite.presents

Allows us to use Jsonite.present and hand it any registered resource to
be presented.

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Unfocus spec

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Track property types when given

Allows for introspection (e.g., for documentation)

We could potentially warn if the value is of the wrong type during
presentation, but this causes complications for, e.g.,
Date/Time/DateTime.

Fixes #12

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Add support for Jsonite(Resource)

It allows us to declare the presented resource class up front:

  class UserPresenter < Jsonite(User)
    property :name
  end
  Jsonite.present(User.first) # => {"name"=>"Stephen"}

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Add mapping cache for resource subclass presentation

E.g.:

  class UserPresenter < Jsonite(User)
    property :name
  end

  class Admin < User
    # ...
  end

  Jsonite.present(Admin.first) # => {"name"=>"Stephen"}

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

kainosnoema commented on lib/jsonite.rb:
```diff
@@ -58,8 +75,8 @@ def present resource, options = {}
-    def property name, options = {}, &handler
-      properties[name.to_s] = { handler: handler }.merge options
+    def property name, type = nil, **options, &handler
+      properties[name.to_s] = { type: type, handler: handler }.merge options
```
Should we document type just a bit?

kainosnoema commented on lib/jsonite.rb:
```diff
@@ -238,3 +259,7 @@ def fetch name, context, options
 
 end
+
+def Jsonite resource_class
```
I'd also like to see a quick example for this one.

stephencelis committed:
Just pass the resource through if no presenter subclass is found

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Document Jsonite() method

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Expose registered resource classes (for introspection)

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Don't bother converting strings

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Document property types

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Spec for metadata

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Add SLOC task

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Jsonite.include_root_in_json support

Uses ActiveRecord's setting by default (if available).

Fixes #4

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Add rehearsal to benchmark spec

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis committed:
Add resource classes to fixtures

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

kainosnoema commented on spec/jsonite_spec.rb:
```diff
@@ -5,29 +5,163 @@
+        it "wraps an array of resources without a class-derivable root key" do
+          presented = presenter.present [resource, resource], root: true
+          expect(presented).to eq(
+            "array"=>[{:name=>"Stephen"}, {:name=>"Stephen"}]
```
This is what I get when presenting in an ActiveRecord::Relation (root key is `array`). I feel like we should be able to determine the root key using `model_name` for relations, no?

kainosnoema replied:
Actually, since we've registered a resource class, I feel like it should be used for the root key here. It's a collection of these resources, so there's no reason we couldn't just use `.model_name.collection` for this. `array` is almost always _not_ what you want.

kainosnoema commented on lib/jsonite.rb:
```diff
@@ -27,23 +49,49 @@ class << self
         resource.present options
-      elsif resource.respond_to?(:to_ary)
+      elsif resource.respond_to? :to_ary
         resource = resource.to_ary.map do |member|
```
The reason I'm getting `array` as the key for ActiveRecord::Relations is because this line turns the resource into an Array, so the call to `Helper.resource_name` below isn't able to determine the correct root key. I'm pretty sure we don't even need to set the `resource` variable here, so removing that should fix things.

stephencelis replied:
Ah, OK. I moved the root key generation above, so this should be fixed.

kainosnoema replied:
Hmm... ya, that seems to work. Any reason why you didn't remove the setting of `resource` here? It's kinda confusing. :/

stephencelis replied:
We're mapping the presentation, so I'm pretty sure we need it?

kainosnoema replied:
We're setting `presented` to the value of the mapped resource, and were only using `resource` for determining the root key. That's why I suggested removing this and leaving the root key generation where it was. Anyway, not a biggie. Just a little confusing.

stephencelis committed:
Calculate root key name before presentation

This way, ActiveRecord::Relation objects will be properly used to
generate root keys.

Signed-off-by: Stephen Celis <stephen@stephencelis.com>

stephencelis commented:
Spring cleaning 📥

kainosnoema commented:
@stephencelis ya, it's been too long. Believe it or not Zinc is still running strong on Jsonite, so we should probably contribute back to this soon and go 1.0 still. Would be good to move into maintenance mode.
